File: tst.py
from sxl.spacetime import *
from sxl.einstein import *
from sympy import *
import sympy

coords = Coordinates("t r z")
t, r, z = coords.coordinate_symbols
c, G, M = symbols("c G M")
a1, b1, a2, b2 = symbols("a1 b1 a2 b2")
sch = 1 - (2*G*M)/(r * c**2)
alpha = Function("alpha")(r)
chi = Function("chi")(r)
metric = MetricTensor([
	[alpha, 0, 0],
	[0, -1/alpha, 0],
	[0, 0, -r**2],
], coords)

manifold = Manifold(metric)

manifold.define(ChristoffelSymbols)
manifold.define(RiemannTensor)
manifold.define(RicciTensor)
manifold.define(RicciScalar)
manifold.define(EinsteinTensor)

ein = manifold.of(EinsteinTensor)

print("\n" * 4)
for i in range(4):
	sympy.pprint(ein.contra(i, i))
	input()
	print("\n" * 4)

# sympy.pprint(ein.contra(0, 0)*metric.co(0, 0) + ein.contra(1, 1)*metric.co(1, 1) + ein.contra(2, 2)*metric.co(2, 2))

# for i in range(4):
# 	for j in range(4):
# 		for k in range(4):
# 			for l in range(4):
# 				print(i, j, k, l)
# 				pprint(manifold.of(RiemannTensor).mixed(i, j, k, l))
# 				print()

File: tst.cpp
#include "sxl-c++/headers/einstein.h"

int main() {
	geometry::CoordinateSystem coords = {"t", "Z", "R", "theta"};
	Symbol c("c");
	Symbol V("V");
	Symbol Z = coords.x(1);
	Symbol R = coords.x(2);
	Symbol k("k");
	Expression rhosq = pow(R, 2) - pow(Z, 2);
	Expression f = 1 - R;
	// c=G=M=1
	geometry::MetricTensor metric({
		{pow(c, 2) - pow(V, 2) + f*k, V, 0, 0},
		{V, -1 - (pow(Z, 2) / rhosq), R * Z / rhosq, 0},
		{0, R * Z / rhosq, -pow(R, 2) / rhosq, 0},
		{0, 0, 0, rhosq}
	}, coords);
	geometry::Manifold mf(metric);
	// tensors::ConnectionCoefficients ccs(metric);
	// mf.define(&ccs);
	// tensors::TorsionTensor tt(metric);
	// mf.define(&tt);
	cout << "ccs" << endl;
	mf.define<tensors::ConnectionCoefficients>();
	cout << "torsion" << endl;
	mf.define<tensors::TorsionTensor>();
	cout << "riemann" << endl;
	mf.define<tensors::RiemannTensor>();
	cout << "ricci" << endl;
	mf.define<tensors::RicciTensor>();
	cout << "einstein" << endl;
	mf.define<einstein::EinsteinTensor>();
	mf.define<einstein::StressEnergyMomentumTensor>();
	mf.define<einstein::LandauLifschitzPseudotensor>();
	std::string _;
	cout << GiNaC::latex;
	// for (int i = 0; i < 4; i++) {
	// 	for (int j = 0; j < 4; j++) {
	// 		for (int k = 0; k < 4; k++) {
	// 			cout << i << j << k << " " << mf.mixed(CCS, {i, j, k}) << endl;
	// 		}
	// 	}
	// }
	// for (int i = 0; i < 4; i++) {
	// 	for (int j = 0; j < 4; j++) {
	// 		for (int k = 0; k < 4; k++) {
	// 			for (int l = 0; l < 4; l++) {
	// 				cout << i+1 << j+1 << k+1 << l+1 << " " << mf.mixed(RIEMANN, {i, j, k, l}) << endl;
	// 			}
	// 		}
	// 	}
	// }
	for (int i = 0; i < 4; i++) {
		for (int j = 0; j < 4; j++) {
			cout << i << j << " " << mf.contra(SEM, {i, j}) << endl;
		}
	}
}